src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,282 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class Binder
+    {
+        /// <summary>
+        /// If the left and right are tuples of matching cardinality, we'll try to bind the operator element-wise.
+        /// When that succeeds, the element-wise conversions are collected and their result type is applied as tuple conversions.
+        /// The element-wise binary operators are collected and stored as a tree for lowering.
+        /// </summary>
+        private BoundTupleBinaryOperator BindTupleBinaryOperator(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            TupleBinaryOperatorInfo operators = BindTupleBinaryOperatorInfo(node, kind, left, right, diagnostics);
+
+            BoundExpression convertedLeft = GenerateConversionForAssignment(operators.LeftConvertedType, left, diagnostics);
+            BoundExpression convertedRight = GenerateConversionForAssignment(operators.RightConvertedType, right, diagnostics);
+
+            TypeSymbol resultType = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
+            return new BoundTupleBinaryOperator(node, convertedLeft, convertedRight, kind, operators, resultType);
+        }
+
+        private TupleBinaryOperatorInfo BindTupleBinaryOperatorInfo(BinaryExpressionSyntax node, BinaryOperatorKind kind,
+            BoundExpression left, BoundExpression right, DiagnosticBag diagnostics)
+        {
+            int? leftCardinality = GetTupleCardinality(left);
+            int? rightCardinality = GetTupleCardinality(right);
+
+            if (leftCardinality.HasValue && rightCardinality.HasValue)
+            {
+                TypeSymbol leftType = left.Type;
+                TypeSymbol rightType = right.Type;
+
+                if (leftCardinality.Value != rightCardinality.Value)
```

Comment:
>leftCardinality.Value != rightCardinality.Value [](start = 20, length = 47)

What if there is a conversion from the left to the right type, or vise versa.

Reply:
Conversions *from* tuple are handled in the regular (non-tuple) case. Conversions *to* tuple do not contribute to this feature.

---
In reply to: [170491495](https://github.com/dotnet/roslyn/pull/23356#discussion_r170491495) [](ancestors = 170491495)